File: deploy.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

// AppliedStateDeltas records live operations that completed before a state
// apply returned. It is suitable for dependency-aware rollback.
type AppliedStateDeltas struct {
	PeerAdds    []WGPeerDeltaOp
	IPSetDeltas []IpsetDeltaOp
	PeerRemoves []WGPeerDeltaOp
}

// ApplyIPSetDeltas applies ipset add/delete operations from deltas through sys.
// Operations are applied in order; the first error encountered is returned.
func ApplyIPSetDeltas(deltas []IpsetDeltaOp, sys SystemAdapter) error {
	_, err := ApplyIPSetDeltasTracked(deltas, sys)
	return err
}

// ApplyIPSetDeltasTracked applies deltas and returns the operations that
// completed before any error. Callers can invert those operations for
// best-effort rollback.
func ApplyIPSetDeltasTracked(deltas []IpsetDeltaOp, sys SystemAdapter) ([]IpsetDeltaOp, error) {
	applied := make([]IpsetDeltaOp, 0, len(deltas))
	for _, d := range deltas {
		if d.Add {
			if err := sys.IPSetAdd(d.Set, d.Entry, d.Comment); err != nil {
				return applied, err
			}
		} else {
			if err := sys.IPSetDel(d.Set, d.Entry); err != nil {
				return applied, err
			}
		}
		applied = append(applied, d)
	}
	return applied, nil
}

// InvertIPSetDeltas returns inverse operations in reverse order for rollback.
func InvertIPSetDeltas(deltas []IpsetDeltaOp) []IpsetDeltaOp {
	inverted := make([]IpsetDeltaOp, 0, len(deltas))
	for i := len(deltas) - 1; i >= 0; i-- {
		d := deltas[i]
		d.Add = !d.Add
		inverted = append(inverted, d)
	}
	return inverted
}

func diffExpectedIPSets(cfg *Config, oldAll, oldMatrix, newAll, newMatrix map[string]string) []IpsetDeltaOp {
	var deltas []IpsetDeltaOp
	deltas = append(deltas, diffExpectedIPSet(cfg.Sets.All, oldAll, newAll)...)
	deltas = append(deltas, diffExpectedIPSet(cfg.Sets.Matrix, oldMatrix, newMatrix)...)
	sort.Slice(deltas, func(i, j int) bool {
		if deltas[i].Set != deltas[j].Set {
			return deltas[i].Set < deltas[j].Set
		}
		if deltas[i].Entry != deltas[j].Entry {
			return deltas[i].Entry < deltas[j].Entry
		}
		return !deltas[i].Add && deltas[j].Add
	})
	return deltas
}

func diffExpectedIPSet(setname string, oldExpected, newExpected map[string]string) []IpsetDeltaOp {
	var deltas []IpsetDeltaOp
	for entry, comment := range newExpected {
		if _, ok := oldExpected[entry]; !ok {
			deltas = append(deltas, IpsetDeltaOp{
				Set:     setname,
				Entry:   entry,
				Comment: comment,
				Add:     true,
			})
		}
	}
	for entry, comment := range oldExpected {
		if _, ok := newExpected[entry]; !ok {
			deltas = append(deltas, IpsetDeltaOp{
				Set:     setname,
				Entry:   entry,
				Comment: comment,
				Add:     false,
			})
		}
	}
	return deltas
}

// ApplyPeerDeltas applies WireGuard peer operations from deltas through sys.
// Operations are applied in order; the first error encountered is returned.
func ApplyPeerDeltas(iface string, deltas []WGPeerDeltaOp, sys SystemAdapter) error {
	_, err := ApplyPeerDeltasTracked(iface, deltas, sys)
	return err
}

// ApplyPeerDeltasTracked applies peer deltas and returns the operations that
// completed before any error. Callers can invert those operations for
// best-effort rollback.
func ApplyPeerDeltasTracked(iface string, deltas []WGPeerDeltaOp, sys SystemAdapter) ([]WGPeerDeltaOp, error) {
	applied := make([]WGPeerDeltaOp, 0, len(deltas))
	for _, d := range deltas {
		switch d.Action {
		case WGPeerAdd:
			if err := sys.WGSetPeer(iface, d.PubKey, d.AllowedIP); err != nil {
				return applied, fmt.Errorf("add WireGuard peer for %q: %w", d.User, err)
			}
		case WGPeerRemove:
			if err := sys.WGDelPeer(iface, d.PubKey); err != nil {
				return applied, fmt.Errorf("remove WireGuard peer for %q: %w", d.User, err)
			}
		default:
			return applied, fmt.Errorf("unknown WireGuard peer delta action %q for %q", d.Action, d.User)
		}
		applied = append(applied, d)
	}
	return applied, nil
}

// InvertPeerDeltas returns inverse peer operations in reverse order for
// rollback.
func InvertPeerDeltas(deltas []WGPeerDeltaOp) []WGPeerDeltaOp {
	inverted := make([]WGPeerDeltaOp, 0, len(deltas))
	for i := len(deltas) - 1; i >= 0; i-- {
		d := deltas[i]
		switch d.Action {
		case WGPeerAdd:
			d.Action = WGPeerRemove
		case WGPeerRemove:
			d.Action = WGPeerAdd
		}
		inverted = append(inverted, d)
	}
	return inverted
}

// ApplyStateDeltas applies WireGuard and ipset operations in a dependency-aware
// order: peer additions, ipset changes, then peer removals.
func ApplyStateDeltas(iface string, ipsetDeltas []IpsetDeltaOp, peerDeltas []WGPeerDeltaOp, sys SystemAdapter) error {
	_, err := ApplyStateDeltasTracked(iface, ipsetDeltas, peerDeltas, sys)
	return err
}

// ApplyStateDeltasTracked applies WireGuard and ipset operations in a
// dependency-aware order and returns the completed operations for rollback.
func ApplyStateDeltasTracked(iface string, ipsetDeltas []IpsetDeltaOp, peerDeltas []WGPeerDeltaOp, sys SystemAdapter) (AppliedStateDeltas, error) {
	var applied AppliedStateDeltas
	var err error

	applied.PeerAdds, err = ApplyPeerDeltasTracked(iface, filterPeerDeltas(peerDeltas, WGPeerAdd), sys)
	if err != nil {
		return applied, err
	}
	applied.IPSetDeltas, err = ApplyIPSetDeltasTracked(ipsetDeltas, sys)
	if err != nil {
		return applied, err
	}
	applied.PeerRemoves, err = ApplyPeerDeltasTracked(iface, filterPeerDeltas(peerDeltas, WGPeerRemove), sys)
	if err != nil {
		return applied, err
	}
	return applied, nil
}

// RollbackStateDeltas rolls back completed state operations in reverse
// dependency order: peer removals, ipset changes, then peer additions.
func RollbackStateDeltas(iface string, applied AppliedStateDeltas, sys SystemAdapter) error {
	var errs []string
	if err := ApplyPeerDeltas(iface, InvertPeerDeltas(applied.PeerRemoves), sys); err != nil {
		errs = append(errs, err.Error())
	}
	if err := ApplyIPSetDeltas(InvertIPSetDeltas(applied.IPSetDeltas), sys); err != nil {
		errs = append(errs, err.Error())
	}
	if err := ApplyPeerDeltas(iface, InvertPeerDeltas(applied.PeerAdds), sys); err != nil {
		errs = append(errs, err.Error())
	}
	if len(errs) > 0 {
		return fmt.Errorf("%s", strings.Join(errs, "; "))
	}
	return nil
}

func filterPeerDeltas(deltas []WGPeerDeltaOp, action WGPeerDeltaAction) []WGPeerDeltaOp {
	out := make([]WGPeerDeltaOp, 0, len(deltas))
	for _, d := range deltas {
		if d.Action == action {
			out = append(out, d)
		}
	}
	return out
}

File: rust-port/src/deploy.rs
use std::collections::HashMap;

use crate::model::{Config, ExpectedIPSets, IPSetDelta, WGPeerDelta, WGPeerDeltaAction};
use crate::system::SystemAdapter;

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct AppliedStateDeltas {
    pub peer_adds: Vec<WGPeerDelta>,
    pub ipset_deltas: Vec<IPSetDelta>,
    pub peer_removes: Vec<WGPeerDelta>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ApplyError<T> {
    pub applied: T,
    pub error: String,
}

pub fn apply_ipset_deltas<S: SystemAdapter>(
    deltas: &[IPSetDelta],
    system: &S,
) -> Result<(), String> {
    apply_ipset_deltas_tracked(deltas, system)
        .map(|_| ())
        .map_err(|err| err.error)
}

pub fn apply_ipset_deltas_tracked<S: SystemAdapter>(
    deltas: &[IPSetDelta],
    system: &S,
) -> Result<Vec<IPSetDelta>, ApplyError<Vec<IPSetDelta>>> {
    let mut applied = Vec::with_capacity(deltas.len());
    for delta in deltas {
        let result = if delta.add {
            system.ipset_add(&delta.set, &delta.entry, &delta.comment)
        } else {
            system.ipset_del(&delta.set, &delta.entry)
        };
        if let Err(error) = result {
            return Err(ApplyError { applied, error });
        }
        applied.push(delta.clone());
    }
    Ok(applied)
}

pub fn invert_ipset_deltas(deltas: &[IPSetDelta]) -> Vec<IPSetDelta> {
    deltas
        .iter()
        .rev()
        .map(|delta| {
            let mut inverted = delta.clone();
            inverted.add = !inverted.add;
            inverted
        })
        .collect()
}

pub fn diff_expected_ipsets(
    cfg: &Config,
    old_expected: &ExpectedIPSets,
    new_expected: &ExpectedIPSets,
) -> Vec<IPSetDelta> {
    let mut deltas = Vec::new();
    deltas.extend(diff_expected_ipset(
        &cfg.sets.all,
        &old_expected.all,
        &new_expected.all,
    ));
    deltas.extend(diff_expected_ipset(
        &cfg.sets.ip_matrix,
        &old_expected.ip_matrix,
        &new_expected.ip_matrix,
    ));
    deltas.extend(diff_expected_ipset(
        &cfg.sets.port_matrix,
        &old_expected.port_matrix,
        &new_expected.port_matrix,
    ));
    deltas.sort_by(|a, b| (&a.set, &a.entry, a.add).cmp(&(&b.set, &b.entry, b.add)));
    deltas
}

pub fn diff_expected_ipset(
    set_name: &str,
    old_expected: &HashMap<String, String>,
    new_expected: &HashMap<String, String>,
) -> Vec<IPSetDelta> {
    let mut deltas = Vec::new();
    for (entry, comment) in new_expected {
        if !old_expected.contains_key(entry) {
            deltas.push(IPSetDelta {
                set: set_name.to_string(),
                entry: entry.clone(),
                comment: comment.clone(),
                add: true,
            });
        }
    }
    for (entry, comment) in old_expected {
        if !new_expected.contains_key(entry) {
            deltas.push(IPSetDelta {
                set: set_name.to_string(),
                entry: entry.clone(),
                comment: comment.clone(),
                add: false,
            });
        }
    }
    deltas
}

pub fn apply_peer_deltas<S: SystemAdapter>(
    iface: &str,
    deltas: &[WGPeerDelta],
    system: &S,
) -> Result<(), String> {
    apply_peer_deltas_tracked(iface, deltas, system)
        .map(|_| ())
        .map_err(|err| err.error)
}

pub fn apply_peer_deltas_tracked<S: SystemAdapter>(
    iface: &str,
    deltas: &[WGPeerDelta],
    system: &S,
) -> Result<Vec<WGPeerDelta>, ApplyError<Vec<WGPeerDelta>>> {
    let mut applied = Vec::with_capacity(deltas.len());
    for delta in deltas {
        let result = match delta.action {
            WGPeerDeltaAction::Add => system
                .wg_set_peer(iface, &delta.pub_key, &delta.allowed_ip)
                .map_err(|err| format!("add WireGuard peer for {:?}: {err}", delta.user)),
            WGPeerDeltaAction::Remove => system
                .wg_del_peer(iface, &delta.pub_key)
                .map_err(|err| format!("remove WireGuard peer for {:?}: {err}", delta.user)),
        };
        if let Err(error) = result {
            return Err(ApplyError { applied, error });
        }
        applied.push(delta.clone());
    }
    Ok(applied)
}

pub fn invert_peer_deltas(deltas: &[WGPeerDelta]) -> Vec<WGPeerDelta> {
    deltas
        .iter()
        .rev()
        .map(|delta| {
            let mut inverted = delta.clone();
            inverted.action = match inverted.action {
                WGPeerDeltaAction::Add => WGPeerDeltaAction::Remove,
                WGPeerDeltaAction::Remove => WGPeerDeltaAction::Add,
            };
            inverted
        })
        .collect()
}

pub fn apply_state_deltas<S: SystemAdapter>(
    iface: &str,
    ipset_deltas: &[IPSetDelta],
    peer_deltas: &[WGPeerDelta],
    system: &S,
) -> Result<(), String> {
    apply_state_deltas_tracked(iface, ipset_deltas, peer_deltas, system)
        .map(|_| ())
        .map_err(|err| err.error)
}

pub fn apply_state_deltas_tracked<S: SystemAdapter>(
    iface: &str,
    ipset_deltas: &[IPSetDelta],
    peer_deltas: &[WGPeerDelta],
    system: &S,
) -> Result<AppliedStateDeltas, ApplyError<AppliedStateDeltas>> {
    let mut applied = AppliedStateDeltas::default();

    let peer_adds = filter_peer_deltas(peer_deltas, WGPeerDeltaAction::Add);
    applied.peer_adds =
        apply_peer_deltas_tracked(iface, &peer_adds, system).map_err(|err| ApplyError {
            applied: AppliedStateDeltas {
                peer_adds: err.applied,
                ..AppliedStateDeltas::default()
            },
            error: err.error,
        })?;

    applied.ipset_deltas =
        apply_ipset_deltas_tracked(ipset_deltas, system).map_err(|err| ApplyError {
            applied: AppliedStateDeltas {
                peer_adds: applied.peer_adds.clone(),
                ipset_deltas: err.applied,
                peer_removes: Vec::new(),
            },
            error: err.error,
        })?;

    let peer_removes = filter_peer_deltas(peer_deltas, WGPeerDeltaAction::Remove);
    applied.peer_removes =
        apply_peer_deltas_tracked(iface, &peer_removes, system).map_err(|err| ApplyError {
            applied: AppliedStateDeltas {
                peer_adds: applied.peer_adds.clone(),
                ipset_deltas: applied.ipset_deltas.clone(),
                peer_removes: err.applied,
            },
            error: err.error,
        })?;

    Ok(applied)
}

pub fn rollback_state_deltas<S: SystemAdapter>(
    iface: &str,
    applied: &AppliedStateDeltas,
    system: &S,
) -> Result<(), String> {
    let mut errors = Vec::new();

    if let Err(err) = apply_peer_deltas(iface, &invert_peer_deltas(&applied.peer_removes), system) {
        errors.push(err);
    }
    if let Err(err) = apply_ipset_deltas(&invert_ipset_deltas(&applied.ipset_deltas), system) {
        errors.push(err);
    }
    if let Err(err) = apply_peer_deltas(iface, &invert_peer_deltas(&applied.peer_adds), system) {
        errors.push(err);
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(errors.join("; "))
    }
}

fn filter_peer_deltas(deltas: &[WGPeerDelta], action: WGPeerDeltaAction) -> Vec<WGPeerDelta> {
    deltas
        .iter()
        .filter(|delta| delta.action == action)
        .cloned()
        .collect()
}
